util/vole.go

```diff
@@ -0,0 +1,297 @@
+	mh "github.com/multiformats/go-multihash"
+)
+
+var bsLog = logging.Logger("bitswap")
+
+// Map of CIDs to be requested and whether they
+// had been answered or not (in each test round).
+var testCIDs map[cid.Cid]bool
+
+func init() {
+	// Generate fixed public peer ID.
+	var err error
+	privKey, _, err = crypto.GenerateKeyPairWithReader(crypto.Ed25519, 2048, new(zeroReader))
+	if err != nil {
+		panic(err)
+	}
+
+	// Number of CIDs requested: one group for valid and one for invalid (nonexistent).
+	CID_NUM := 2 //fixme: parameter
+	testCIDs = createCids(CID_NUM)
+}
+
+// FIXME: There is probably a much easier way to get unlimited zero bytes.
+var privKey crypto.PrivKey
+type zeroReader struct {
+}
+func (*zeroReader) Read(p []byte) (n int, err error) {
+	p[0] = 0
+	return 1, nil
+}
+// FIXME: Rename.
+func CheckBitswapCID(ctx context.Context,
+	targetPeer peer.ID) (map[cid.Cid]bool, error) {
+
+	rcv := &bsReceiver{
+		target: targetPeer,
+		result: make(chan msgOrErr),
+	}
+	bs, err := connectToBitSwap(ctx, targetPeer, rcv)
+	if err != nil {
+		return nil, err
+	}
+
+	msg := createMsg(testCIDs)
+
+	if err := bs.SendMessage(ctx, targetPeer, msg); err != nil {
+		return nil, err
+	}
+
+	return trackResponses(ctx, rcv, testCIDs)
+}
+
+// We will wait for the responses as long as the passed context dictates.
+// FIXME: We could also set another timeout here but centralize in caller
+//  for now for simplicity.
+func trackResponses(ctx context.Context, rcv *bsReceiver, requestedCids map[cid.Cid]bool) (map[cid.Cid]bool, error) {
+	for c, _ := range requestedCids {
+		requestedCids[c] = false
+	}
+
+	// Keep track of responded CIDs to exit early.
+	responded := 0
+	markResponse := func(c cid.Cid) {
+		if requestedCids[c] == false {
+			requestedCids[c] = true
+			responded += 1
+		}
+	}
+
+loop:
+	for {
+		var res msgOrErr
+		select {
+		case res = <-rcv.result:
+		case <-ctx.Done():
+			break loop
+		}
+
+		if res.err != nil {
+			return nil, res.err
+		}
+
+		if res.msg == nil {
+			panic("should not be reachable")
+		}
+
+		for _, block := range res.msg.Blocks() {
+			markResponse(block.Cid())
+		}
+
+		for _, c := range res.msg.Haves() {
+			markResponse(c)
+		}
+
+		for _, c := range res.msg.DontHaves() {
+			markResponse(c)
+		}
+		if responded == len(requestedCids) {
+			return nil, nil
+		}
+	}
+
+	missingCids := make(map[cid.Cid]bool)
+	for c, responded := range requestedCids {
+		if !responded {
+			missingCids[c] = true
+		}
+	}
+	return missingCids, nil
+}
+
+func connectToBitSwap(ctx context.Context, targetPeer peer.ID, rcv *bsReceiver) (bsnet.BitSwapNetwork, error) {
+	// Connect randomly on either TCP or QUIC.
+	localhost := ""
+	if rand.Intn(2) == 1 {
+		localhost = "/ip4/127.0.0.1/udp/4001/quic/p2p/" + targetPeer.String()
+		bsLog.Infof("connecting through QUIC")
+	} else {
+		localhost = "/ip4/127.0.0.1/tcp/4001/p2p/" + targetPeer.String()
+		bsLog.Infof("connecting through TCP")
+	}
+	// FIXME: Any preference on either transport?
```

Comment:
Ideally, we'd try both. We don't know if it's a problem here or in the transport.